stdlib/public/core/StringHashable.swift

```diff
@@ -79,19 +92,48 @@ extension _StringGutsSlice {
     var index = self.range.lowerBound
     let cachedEndIndex = self.range.upperBound
     
-    let normalize: (String.Index, _StringGuts, UnsafeMutableBufferPointer<UInt8>,
-      UnsafeMutableBufferPointer<UInt16>, UnsafeMutableBufferPointer<UInt16>
-    ) -> NormalizationResult
-    if _fastPath(self.isFastUTF8) {
-      normalize = fastNormalize
-    } else {
-      normalize = foreignNormalize
+    var bufferCleanup: (() -> ())? = nil
+    while index != cachedEndIndex {
+      let result = _fastNormalize(index, sourceBuffer, outputBuffer, icuInputBuffer, icuOutputBuffer)
+      switch result {
+      case let .success(r):
+        for i in 0..<r.amountFilled {
+          hasher.combine(outputBuffer[i])
+        }
+        index = r.nextReadPosition
+      case let .bufferTooSmall(resize):
+        _internalInvariant(bufferCleanup == nil)
+        outputBuffer = UnsafeMutableBufferPointer<UInt8>.allocate(capacity: resize.output)
+        icuInputBuffer = UnsafeMutableBufferPointer<UInt16>.allocate(capacity: resize.icuInput)
+        icuOutputBuffer = UnsafeMutableBufferPointer<UInt16>.allocate(capacity: resize.icuOutput)
+        bufferCleanup = {
+          outputBuffer.deallocate()
+          icuInputBuffer.deallocate()
+          icuOutputBuffer.deallocate()
+        }
+      }
+    }
+    if let cleanup = bufferCleanup {
+      cleanup()
     }
+  }
+  
```

Comment:
Seems like all this logic could be refactored into a helper function that took the buffers `inout` and told the caller if they're responsible for calling `deallocate`.